// sanity-plugin-media-library/lib/components/LabelWithInput.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LabelWithInput = void 0;
var react_1 = __importDefault(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var StyledContainer = styled_components_1.default.label(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  cursor: pointer;\n  display: block;\n  width: 100%;\n"], ["\n  cursor: pointer;\n  display: block;\n  width: 100%;\n"])));
var StyledWrapper = styled_components_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  align-items: center;\n  cursor: pointer;\n  display: flex;\n  margin-bottom: 8px;\n  width: 100%;\n\n  & > input {\n    margin-left: 16px;\n  }\n\n  & > span {\n    font-size: 14px;\n    margin: 0;\n    opacity: 0.8;\n  }\n"], ["\n  align-items: center;\n  cursor: pointer;\n  display: flex;\n  margin-bottom: 8px;\n  width: 100%;\n\n  & > input {\n    margin-left: 16px;\n  }\n\n  & > span {\n    font-size: 14px;\n    margin: 0;\n    opacity: 0.8;\n  }\n"])));
var StyledLabel = styled_components_1.default.span(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  color: ", ";\n  display: block;\n  font-family: ", ";\n  font-size: 16px;\n  font-weight: 500;\n  line-height: 1.1;\n  margin: 0 0 0.5em;\n  width: 100%;\n"], ["\n  color: ", ";\n  display: block;\n  font-family: ", ";\n  font-size: 16px;\n  font-weight: 500;\n  line-height: 1.1;\n  margin: 0 0 0.5em;\n  width: 100%;\n"])), function (_a) {
    var theme = _a.theme;
    return theme.inputLabelColor;
}, function (_a) {
    var theme = _a.theme;
    return theme.appFontFamily;
});
var StyledInput = styled_components_1.default.input(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  background-color: ", ";\n  border-radius: ", ";\n  border: 0;\n  color: ", ";\n  font-family: ", ";\n  font-size: 16px;\n  line-height: 1.1;\n  outline: 0;\n  padding: 16px;\n  width: 100%;\n"], ["\n  background-color: ", ";\n  border-radius: ", ";\n  border: 0;\n  color: ", ";\n  font-family: ", ";\n  font-size: 16px;\n  line-height: 1.1;\n  outline: 0;\n  padding: 16px;\n  width: 100%;\n"])), function (_a) {
    var theme = _a.theme;
    return theme.inputBackgroundColor;
}, function (_a) {
    var theme = _a.theme;
    return theme.appBorderRadius;
}, function (_a) {
    var theme = _a.theme;
    return theme.inputTextColor;
}, function (_a) {
    var theme = _a.theme;
    return theme.appFontFamily;
});
var StyledTextArea = styled_components_1.default.textarea(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  background-color: ", ";\n  border-radius: ", ";\n  border: 0;\n  color: ", ";\n  font-family: ", ";\n  font-size: 16px;\n  line-height: 1.1;\n  outline: 0;\n  padding: 16px;\n  resize: vertical;\n  rows: 2;\n  width: 100%;\n"], ["\n  background-color: ", ";\n  border-radius: ", ";\n  border: 0;\n  color: ", ";\n  font-family: ", ";\n  font-size: 16px;\n  line-height: 1.1;\n  outline: 0;\n  padding: 16px;\n  resize: vertical;\n  rows: 2;\n  width: 100%;\n"])), function (_a) {
    var theme = _a.theme;
    return theme.inputBackgroundColor;
}, function (_a) {
    var theme = _a.theme;
    return theme.appBorderRadius;
}, function (_a) {
    var theme = _a.theme;
    return theme.inputTextColor;
}, function (_a) {
    var theme = _a.theme;
    return theme.appFontFamily;
});
var StyledSelectWrapper = styled_components_1.default.div(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  position: relative;\n"], ["\n  position: relative;\n"])));
var StyledSelectArrow = styled_components_1.default.span(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  background-color: ", ";\n  bottom: 0;\n  clip-path: polygon(100% 0%, 0 0%, 50% 100%);\n  content: '';\n  height: 0.5em;\n  pointer-events: none;\n  position: absolute;\n  right: 16px;\n  top: 50%;\n  transform: translateY(-50%);\n  width: 0.8em;\n"], ["\n  background-color: ", ";\n  bottom: 0;\n  clip-path: polygon(100% 0%, 0 0%, 50% 100%);\n  content: '';\n  height: 0.5em;\n  pointer-events: none;\n  position: absolute;\n  right: 16px;\n  top: 50%;\n  transform: translateY(-50%);\n  width: 0.8em;\n"])), function (_a) {
    var theme = _a.theme;
    return theme.inputTextColor;
});
var StyledSelect = styled_components_1.default.select(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  appearance: none;\n  background-color: ", ";\n  border-radius: ", ";\n  border: 0;\n  color: ", ";\n  font-family: ", ";\n  font-size: 16px 16px 34px 16px;\n  line-height: 1.1;\n  outline: 0;\n  padding: 16px;\n  width: 100%;\n"], ["\n  appearance: none;\n  background-color: ", ";\n  border-radius: ", ";\n  border: 0;\n  color: ", ";\n  font-family: ", ";\n  font-size: 16px 16px 34px 16px;\n  line-height: 1.1;\n  outline: 0;\n  padding: 16px;\n  width: 100%;\n"])), function (_a) {
    var theme = _a.theme;
    return theme.inputBackgroundColor;
}, function (_a) {
    var theme = _a.theme;
    return theme.appBorderRadius;
}, function (_a) {
    var theme = _a.theme;
    return theme.inputTextColor;
}, function (_a) {
    var theme = _a.theme;
    return theme.appFontFamily;
});
var LabelWithInput = function (_a) {
    var label = _a.label, onChange = _a.onChange, _b = _a.type, type = _b === void 0 ? 'text' : _b, rest = __rest(_a, ["label", "onChange", "type"]);
    var elementsMap = {
        checkbox: Checkbox,
        location: Location,
        textArea: StyledTextArea,
        select: Select,
    };
    var onChangeMap = {
        checkbox: function (e) { return onChange(e.target.checked); },
        location: function (value) { return onChange(value); },
        number: function (e) { return onChange(parseFloat(e.target.value)); },
        select: function (e) { return onChange(e.target.value); },
    };
    var defaultOnChange = function (e) { return onChange(e.target.value); };
    var handleOnChange = onChangeMap[type] || defaultOnChange;
    var Element = elementsMap[type] || StyledInput;
    return (react_1.default.createElement(StyledContainer, null,
        react_1.default.createElement(StyledLabel, null, label),
        react_1.default.createElement(Element, __assign({ onChange: handleOnChange, type: type }, rest))));
};
exports.LabelWithInput = LabelWithInput;
var Checkbox = function (_a) {
    var value = _a.value, rest = __rest(_a, ["value"]);
    return react_1.default.createElement("input", __assign({ checked: !!value }, rest));
};
var Location = function (_a) {
    var onChange = _a.onChange, _b = _a.value, value = _b === void 0 ? {} : _b, rest = __rest(_a, ["onChange", "value"]);
    var fields = [
        { label: 'Latitude', name: 'lat', type: 'number', step: 'any' },
        { label: 'Longitude', name: 'lng', type: 'number', step: 'any' },
        { label: 'Altitude', name: 'alt', type: 'number', step: 'any' },
    ];
    var handleInputchange = function (e) {
        var _a;
        onChange(__assign(__assign({}, (value || {})), (_a = {}, _a[e.target.name] = parseFloat(e.target.value), _a)));
    };
    return (react_1.default.createElement(StyledContainer, __assign({ as: "div" }, rest), fields.map(function (_a) {
        var label = _a.label, name = _a.name, rest = __rest(_a, ["label", "name"]);
        return (react_1.default.createElement(StyledWrapper, { key: name },
            react_1.default.createElement(StyledLabel, null, label),
            react_1.default.createElement(StyledInput, __assign({ onChange: handleInputchange, name: name, value: value[name] }, rest))));
    })));
};
var Select = function (_a) {
    var options = _a.options, placeholder = _a.placeholder, value = _a.value, rest = __rest(_a, ["options", "placeholder", "value"]);
    return (react_1.default.createElement(StyledSelectWrapper, null,
        react_1.default.createElement(StyledSelect, __assign({}, rest),
            react_1.default.createElement("option", { value: "" }, placeholder), options === null || options === void 0 ? void 0 :
            options.map(function (option) { return (react_1.default.createElement("option", { key: option.value, value: option.value, selected: option.value === value }, option.title)); })),
        react_1.default.createElement(StyledSelectArrow, null)));
};
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8;
