// sanity-plugin-vercel-deploy/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _vercelDeploy = _interopRequireDefault(require("./vercel-deploy"));

var deployIcon = function deployIcon() {
  return /*#__PURE__*/_react["default"].createElement("svg", {
    "data-sanity-icon": "true",
    viewBox: "0 0 25 25",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    preserveAspectRatio: "xMidYMid",
    width: "1em",
    height: "1em"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M10.06,19.53H4.27L12.5,5.21l8.23,14.32H14.94",
    style: {
      stroke: 'currentColor',
      strokeWidth: '1.2'
    }
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M12.5,12.58v7.51",
    style: {
      stroke: 'currentColor',
      strokeWidth: '1.2'
    }
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M15.12,16.76,12.5,12.58,10,16.71",
    style: {
      stroke: 'currentColor',
      strokeWidth: '1.2'
    }
  }));
};

var _default = {
  title: 'Deploy',
  name: 'vercel-deploy',
  icon: deployIcon,
  component: _vercelDeploy["default"]
};
exports["default"] = _default;
