// sanity-plugin-media-library/lib/themes/darkTheme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.darkTheme = void 0;
var colorBlack = '#000';
var colorWhite = '#fff';
var colorOffBlack = '#111';
var colorBorder = '#222';
var colorBlackLight = '#333';
var colorGray = '#666';
var colorPrimary = '#FFE900';
var colorError = '#FF1700';
exports.darkTheme = {
    appBackgroundColor: colorBlack,
    appBorderRadius: '2px',
    appTextColor: colorWhite,
    appFontFamily: "system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', sans-serif",
    assetModalBorderColor: colorBorder,
    assetModalInfoTextColor: colorGray,
    assetModalInfoTitleColor: colorWhite,
    bottomBarBorderColor: colorBorder,
    buttonPrimaryBackgroundColor: colorPrimary,
    buttonPrimaryBorderColor: colorPrimary,
    buttonPrimaryColor: colorBlack,
    buttonSecondaryBackgroundColor: 'transparent',
    buttonSecondaryBorderColor: colorGray,
    buttonSecondaryColor: colorGray,
    draggableMediaItemAmountBackgroundColor: colorPrimary,
    draggableMediaItemAmountColor: colorBlack,
    draggableMediaItemBackgroundColor: colorBlack,
    draggableMediaItemBorderColor: colorPrimary,
    errorNotifcationTextColor: colorWhite,
    errorNotificationBackgroundColor: colorError,
    errorNotificationCloseColor: colorWhite,
    filterListActiveColor: colorWhite,
    filterListInactiveColor: colorGray,
    inputBackgroundColor: colorBlackLight,
    inputIconColor: colorGray,
    inputLabelColor: colorWhite,
    inputTextColor: colorWhite,
    loaderColor: colorWhite,
    mediaGridSelectedBorderColor: colorPrimary,
    mediaItemBackgroundColor: colorBlackLight,
    mediaItemIconColor: colorGray,
    mediaItemTextColor: colorWhite,
    mediaLibraryBackgroundColor: colorOffBlack,
    mediaListBorderColor: colorBorder,
    mediaListSelectedBorderColor: colorPrimary,
    modalBackgroundColor: colorOffBlack,
    modalBoxShadowColor: colorBlack,
    modalOverlayBackgroundColor: 'rgba(0, 0, 0, 0.5)',
    sidebarBackgroundColor: colorBlack,
    topBarBorderColor: colorBorder,
    topBarButtonActiveColor: colorWhite,
    topBarButtonInactiveColor: colorGray,
    unusedLabelBackgroundColor: colorError,
    unusedLabelTextColor: colorWhite,
    uploadDropAreaBackgroundColor: 'rgba(0, 0, 0, 0.8)',
    uploadDropAreaIconColor: colorWhite,
    usedLabelBackgroundColor: colorBlack,
    usedLabelTextColor: colorGray,
};
