// sanity-plugin-media-library/lib/themes/lightTheme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lightTheme = void 0;
var colorBlack = '#000';
var colorBorder = '#eee';
var colorError = '#FF1700';
var colorGray = '#333';
var colorGrayLight = '#666';
var colorOffBlack = '#000';
var colorOffWhite = '#fcfcfc';
var colorPrimary = '#2276fc';
var colorWhite = '#fff';
exports.lightTheme = {
    appBackgroundColor: colorWhite,
    appBorderRadius: '2px',
    appFontFamily: "system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', sans-serif",
    appTextColor: colorOffBlack,
    assetModalBorderColor: colorBorder,
    assetModalInfoTextColor: colorGray,
    assetModalInfoTitleColor: colorOffBlack,
    bottomBarBorderColor: colorBorder,
    buttonPrimaryBackgroundColor: colorPrimary,
    buttonPrimaryBorderColor: colorPrimary,
    buttonPrimaryColor: colorWhite,
    buttonSecondaryBackgroundColor: 'transparent',
    buttonSecondaryBorderColor: colorPrimary,
    buttonSecondaryColor: colorPrimary,
    draggableMediaItemAmountBackgroundColor: colorPrimary,
    draggableMediaItemAmountColor: colorBlack,
    draggableMediaItemBackgroundColor: colorBlack,
    draggableMediaItemBorderColor: colorPrimary,
    errorNotifcationTextColor: colorWhite,
    errorNotificationBackgroundColor: colorError,
    errorNotificationCloseColor: colorWhite,
    filterListActiveColor: colorOffBlack,
    filterListInactiveColor: colorGrayLight,
    inputBackgroundColor: colorBorder,
    inputIconColor: colorGrayLight,
    inputLabelColor: colorOffBlack,
    inputTextColor: colorOffBlack,
    loaderColor: colorOffBlack,
    mediaGridSelectedBorderColor: colorPrimary,
    mediaItemBackgroundColor: colorBorder,
    mediaItemIconColor: colorGray,
    mediaItemTextColor: colorOffBlack,
    mediaLibraryBackgroundColor: colorOffWhite,
    mediaListBorderColor: colorBorder,
    mediaListSelectedBorderColor: colorPrimary,
    modalBackgroundColor: colorWhite,
    modalBoxShadowColor: 'rgba(0, 0, 0, 0.5)',
    modalOverlayBackgroundColor: 'rgba(0, 0, 0, 0.5)',
    sidebarBackgroundColor: colorWhite,
    topBarBorderColor: colorBorder,
    topBarButtonActiveColor: colorBlack,
    topBarButtonInactiveColor: colorGrayLight,
    unusedLabelBackgroundColor: colorError,
    unusedLabelTextColor: colorWhite,
    uploadDropAreaBackgroundColor: 'rgba(0, 0, 0, 0.8)',
    uploadDropAreaIconColor: colorWhite,
    usedLabelBackgroundColor: colorWhite,
    usedLabelTextColor: colorGray,
};
